### ISSUE TITLE
The interpreter for pick completely disregards the UB precondition

### ISSUE THREAD

**Issue by YaZko** [vellvm/vellvm#111]
Pick events carry a predicate that the model checks to decide whether the instruction should raise UB, or collect the semantics of all possible concretizations.

The interpreter however (concretize_picks) simply disregards this proposition. 
This leads to some programs that should raise UB to actually run to completion (see #106 for a contrived example).

Three possibilities:
* Possibly, this is an acceptable situation
* We could make the interpreter overly conservative by raising UB in `concretize_picks` as soon as the uvalue is not the trivial embedding of a dvalue
* We could provide a third argument to pick that would be a decidable check that `concretize_picks` would run before deciding whether it should raise UB or not.

**Comment by Chobbes** [comment]
Currently my understanding is that "this is an acceptable situation". If it's UB the interpreter can behave however it wants, so disregarding the predicate is perfectly fine semantically. This seems like a similar situation to concretizing `undef` to 0 --- it results in a trace which is a refinement of the model.

It might be nice to have a version of the interpreter which attempts to check that these predicates hold (this will also have to know something about the memory model?), so we can have a kind of UB checker. Maybe the second possibility is a useful version of this in some cases, but it seems much too aggressive, and this would not be a refinement of our model, right?

**Comment by YaZko** [comment]
Yes I tend to agree with you on all accounts. It is a legitimate behaviour for the interpreter, but I think it should be documented though if we keep it as is as it's not completely intuitive. And there could be use for interpreters behaving differently in this regard.

**Comment by Chobbes** [comment]
For sure. I have never been really happy with the UB events and the predicates in the prop monad being as separated as they are. Conceptually it would be really nice to have UB happen exactly when a UB event is raised... But I'm not really sure how to do that.

**Comment by YaZko** [comment]
Closed as obsolete.
